### Basic Programs

### Question1 - Python Program to Add Two Numbers

- Given two numbers num1 and num2. The task is to write a Python program to find the addition of these two numbers. 

In [3]:
#Here num1 and num2 is a variable and we are going to add the both variable with + operator.

In [4]:
#Operator - https://www.geeksforgeeks.org/python-operators/

In [6]:
#simple program to add two value
num1 = 15
num2 = 12

#adding two nos
sum = num1 + num2

#priting values
print("Sum of", num1 ,"and", num2, "is", sum)

Sum of 15 and 12 is 27


In [7]:
#Adding two numbers with user input

In [8]:
number1 = input("first number: ")
number2 = input("Second number: ")

#Adding two numbers
# User might also enter float numbers

sum = float(number1) + float(number2)

#disply the sum
# Will print value in float

print("The sum of {0} and {1} is {2}".format(number1, number2,sum))

first number: 2
Second number: 2
The sum of 2 and 2 is 4.0


In [9]:
#Defining add function and returning the result

In [10]:
# To define a function that take two integers
# and return the sum of those two numbers

def add(a,b):
    return a+b

#initializing the variables
num1 = 10
num2 = 5

#function calling and store the result into sum_of_twonumbers
sum_of_twonumbers =  add(num1, num2)

#To print the result
print("Sum of {0} and {1} is {2}:".format(num1, num2, sum_of_twonumbers))

Sum of 10 and 5 is 15:


Add two numbers in Python using operator.add() method

In [11]:
# Operator.add() - https://www.geeksforgeeks.org/operator-functions-in-python-set-1/

In [12]:
num1 = 15
num2 = 12

#Adding two nos
import operator
su = operator.add(num1,num2)

#printing values
print("Sum of {0} and {1} is {2}".format(num1, num2,su))

Sum of 15 and 12 is 27


Adding two number using lambda function

In [13]:
# lambda function - https://www.geeksforgeeks.org/python-lambda-anonymous-functions-filter-map-reduce/

In [14]:
#Define a lambda function to add two numbers
add_numbers = lambda x,y : x+y

#Take input from the user
num1 = 1
num2 = 2

#Call the lambda function to add the two numbers
result = add_numbers(num1, num2)

#Print the result
print("The sum of", num1, "and", num2, "is", result)

The sum of 1 and 2 is 3


Python program to add two numbers with recursive function

- This code defines a recursive function to add two numbers by incrementing one number and decrementing the other. User inputs two numbers, and the sum is calculated and displayed

In [ ]:
# Define a recursive function to add two numbers
def add_numbers_recursive(x, y):
    if y == 0:
        return x
    else:
        return add_numbers_recursive(x + 1, y - 1)
 
# Take input from the user
num1 = 1
num2 = 2
 
# Call the recursive function to add the two numbers
result = add_numbers_recursive(num1, num2)
 
# Print the result
print("The sum of", num1, "and", num2, "is", result)

## Question 2. Find Maximum of two numbers in Python

- This is the naive approach where we will compare two numbers using if-else statement and will print the output accordingly.

In [15]:
#IF Else - https://www.geeksforgeeks.org/python-if-else/

In [16]:
# Python program to find the
# maximum of two numbers
 
 
def maximum(a, b):
     
    if a >= b:
        return a
    else:
        return b
     
# Driver code
a = 2
b = 4
print(maximum(a, b))

4


Find Maximum of two numbers Using max() function

In [17]:
#This function is used to find the maximum of the values passed as its arguments.

In [18]:
# Python program to find the
# maximum of two numbers
 
 
a = 2
b = 4
 
maximum = max(a, b)
print(maximum)

4


Maximum of two numbers Using Ternary Operator

In [19]:
#Ternary Oprator - https://www.geeksforgeeks.org/python-operators/

- This operator is also known as conditional expression are operators that evaluate something based on a condition being true or false. It simply allows testing a condition in a single line

In [20]:
# Python program to find the
# maximum of two numbers
     
# Driver code
a = 2
b = 4
 
# Use of ternary operator
print(a if a >= b else b)

4


Maximum of two numbers Using lambda function 

In [22]:
# python code to find maximum of two numbers
 
a=2
b=4
maximum = lambda a,b:a if a > b else b
print(f'{maximum(a,b)} is a maximum number')

4 is a maximum number


Maximum of two numbers Using list comprehension

In [24]:
a=2
b=4

x=[a if a>b else b]
print("maximum number is:",x)

maximum number is: [4]


Maximum of two numbers Using sort() method

In [25]:
# Python program to find the
# maximum of two numbers
a = 2
b = 4
x=[a,b]
x.sort()
print(x[-1])

4


## Question 3 . Python Program to Find the Factorial of a Number

Factorial of a non-negative integer, is multiplication of all integers smaller than or equal to n. 

Find the Factorial of a Number Using Recursive approach

- This Python program uses a recursive function to calculate the factorial of a given number. The factorial is computed by multiplying the number with the factorial of its preceding number.

In [26]:
# Python 3 program to find
# factorial of given number
def factorial(n):
     
    # single line to find factorial
    return 1 if (n==1 or n==0) else n * factorial(n - 1)
 
# Driver Code
num = 5
print("Factorial of",num,"is",factorial(num))

Factorial of 5 is 120


Find the Factorial of a Number Using Iterative approach

In [27]:
# Python 3 program to find 
# factorial of given number
def factorial(n):
    if n < 0:
        return 0
    elif n == 0 or n == 1:
        return 1
    else:
        fact = 1
        while(n > 1):
            fact *= n
            n -= 1
        return fact
 
# Driver Code
num = 5
print("Factorial of",num,"is",
factorial(num))

Factorial of 5 is 120


In [28]:
# Python 3 program to find
# factorial of given number
  
# Function to find factorial of given number
def factorial(n):
       
    res = 1
      
    for i in range(2, n+1):
        res *= i
    return res
 # Driver Code
num = 5
print("Factorial of", num, "is",
factorial(num))

Factorial of 5 is 120


Find the Factorial of a Number Using One line Solution (Using Ternary operator): 

In [29]:
# Python 3 program to find
# factorial of given number
 
def factorial(n):
 
    # single line to find factorial
    return 1 if (n==1 or n==0) else n * factorial(n - 1)
 
 
# Driver Code
num = 5
print ("Factorial of",num,"is",
      factorial(num))

Factorial of 5 is 120


Find the Factorial of a Number Using using In-built function 

In Python, math module contains a number of mathematical operations, which can be performed with ease using the module. math.factorial() function returns the factorial of desired number.

In [31]:
# Python 3 program to find
# factorial of given number
import math
 
def factorial(n):
    return(math.factorial(n))
 
 
# Driver Code
num = 5
print("Factorial of", num, "is",
      factorial(num))

Factorial of 5 is 120


Find the Factorial of a Number Using numpy.prod 

In [32]:
import numpy
n=5
x=numpy.prod([i for i in range(1,n+1)])
print(x)

120


In [34]:
# Python 3 program to find factorial
# of a given number using prime
# factorization method.
 
# Function to find prime factors of a number
def primeFactors(n):
    factors = {}
    i = 2
    while i*i <= n:
        while n % i == 0:
            if i not in factors:
                factors[i] = 0
            factors[i] += 1
            n //= i
        i += 1
    if n > 1:
        if n not in factors:
            factors[n] = 0
        factors[n] += 1
    return factors
 
# Function to find factorial of a number
def factorial(n):
    result = 1
    for i in range(2, n+1):
        factors = primeFactors(i)
        for p in factors:
            result *= p ** factors[p]
    return result
 
# Driver Code
num = 5
print("Factorial of", num, "is", factorial(num))

Factorial of 5 is 120


## Question 4. Python Program for Simple Interest

Simple interest formula is given by: Simple Interest = (P x T x R)/100 Where, P is the principal amount T is the time and R is the rate

Python Program for simple interest

In [35]:
# Python3 program to find simple interest
# for given principal amount, time and
# rate of interest.
 
 
def simple_interest(p,t,r):
    print('The principal is', p)
    print('The time period is', t)
    print('The rate of interest is',r)
     
    si = (p * t * r)/100
     
    print('The Simple Interest is', si)
    return si
     
# Driver code
simple_interest(8, 6, 8)

The principal is 8
The time period is 6
The rate of interest is 8
The Simple Interest is 3.84


3.84

Program for simple interest with Taking input from user

In [36]:
# Python3 program to find simple interest
#  principal amount, time and
# rate of interest taken from user.
 
 
def simple_interest(p,t,r):
    print('The principal is', p)
    print('The time period is', t)
    print('The rate of interest is',r)
     
    si = (p * t * r)/100
     
    print('The Simple Interest is', si)
     
     
# Driver code
P = int(input("Enter the principal amount :"))
T = int(input("Enter the time period :"))
R = int(input("Enter the rate of interest :"))
simple_interest(P,T,R)

Enter the principal amount :2
Enter the time period :3
Enter the rate of interest :3
The principal is 2
The time period is 3
The rate of interest is 3
The Simple Interest is 0.18


## Question 5. Python Program for Compound Interest

In [37]:
# Python3 program to find compound
# interest for given values.
 
 
def compound_interest(principal, rate, time):
 
    # Calculates compound interest
    Amount = principal * (pow((1 + rate / 100), time))
    CI = Amount - principal
    print("Compound interest is", CI)
 
 
# Driver Code
compound_interest(10000, 10.25, 5)

Compound interest is 6288.946267774416


Compound Interest with Input taking from user

In [38]:
# Python3 program to find compound
# interest for input taking from user.
 
 
def compound_interest(principal, rate, time):
 
    # Calculates compound interest
    Amount = principal * (pow((1 + rate / 100), time))
    CI = Amount - principal
    print("Compound interest is", CI)
 
 
# Driver Code
#Taking input from user.
principal = int(input("Enter the principal amount: "))
rate = int(input("Enter rate of interest: "))
time = int(input("Enter time in years: " ))
#Function Call
compound_interest(principal,rate,time)

Enter the principal amount: 3000
Enter rate of interest: 5
Enter time in years: 3
Compound interest is 472.87500000000045


Finding compound interest of given values without using pow() function.

In [39]:
# Python code
# To find compound interest
 
# inputs
p= 1200   # principal amount
t= 2      # time
r= 5.4    # rate
# calculates the compound interest
a=p*(1+(r/100))**t  # formula for calculating amount
ci=a-p  # compound interest = amount - principal amount
# printing compound interest value
print(ci)

133.0992000000001


Compound Interest using for loop

In [40]:
def compound_interest(principal, rate, time):
    Amount = principal
    for i in range(time):
        Amount = Amount * (1 + rate/100)
    CI = Amount - principal
    print("Compound interest is", CI)
# Driver Code
compound_interest(1200, 5.4, 2)

Compound interest is 133.0992000000001


## Queston 6. Python Program to Check Armstrong Number

Given a number x, determine whether the given number is Armstrong number or not. A positive integer of n digits is called an Armstrong number of order n (order is number of digits) if.

abcd... = pow(a,n) + pow(b,n) + pow(c,n) + pow(d,n) + .... 

In [42]:
# Python program to determine whether
# the number is Armstrong number or not
 
# Function to calculate x raised to
# the power y
def power(x, y):
     
    if y == 0:
        return 1
    if y % 2 == 0:
        return power(x, y // 2) * power(x, y // 2)
         
    return x * power(x, y // 2) * power(x, y // 2)
 
# Function to calculate order of the number
def order(x):
 
    # Variable to store of the number
    n = 0
    while (x != 0):
        n = n + 1
        x = x // 10
         
    return n
# Function to check whether the given
# number is Armstrong number or not
def isArmstrong(x):
     
    n = order(x)
    temp = x
    sum1 = 0
     
    while (temp != 0):
        r = temp % 10
        sum1 = sum1 + power(r, n)
        temp = temp // 10
 
    # If condition satisfies
    return (sum1 == x)
 
# Driver code
x = 153
print(isArmstrong(x))
 
x = 1253
print(isArmstrong(x))

True
False


Python program to check if a number is an Armstrong number without using the power function

In [43]:
# python 3 program
# to check whether the given number is armstrong or not
# without using power function
 
n = 153  # or n=int(input()) -> taking input from user
s = n  # assigning input value to the s variable
b = len(str(n))
sum1 = 0
while n != 0:
    r = n % 10
    sum1 = sum1+(r**b)
    n = n//10
if s == sum1:
    print("The given number", s, "is armstrong number")
else:
    print("The given number", s, "is not armstrong number")
 


The given number 153 is armstrong number


Python program to check if a number is an Armstrong number Using string manipulation

This approach involves converting the input number into a string and iterating through each digit in the string. For each digit, we raise it to the power of the number of digits in the input number, and sum up the results. If the final sum equals the input number, it is an Armstrong number.

Algorithm
1. Convert the input number into a string using str(num).
2. Find the length of the string using len(num_str) and store it in n.
3. Initialize a variable sum to zero.
4. Iterate through each digit in the string using a for loop, and convert each digit back to an integer using int(digit).
5. Raise each digit to the power of n using int(digit)**n, and add the result to sum.
6. After the loop is complete, check whether sum is equal to num.
7. If sum is equal to num, return True (the input number is an Armstrong number).
8. If sum is not equal to num, return False (the input number is not an Armstrong number).

In [44]:
def is_armstrong(num):
    num_str = str(num)
    n = len(num_str)
    sum = 0
    for digit in num_str:
        sum += int(digit)**n
    if sum == num:
        return True
    else:
        return False
num=153
print(is_armstrong(num))

True


Digit by digit sum approach to check for Armstrong number

1. Find the number of digits in the given number.
2. Initialize a variable sum to zero.
3. Extract each digit from the given number and raise it to the power of the number of digits and add it to the sum.
4. If the sum is equal to the given number, then it is an Armstrong number, else it is not. #include <iostream>

In [45]:
import math
 
def isArmstrong(num):
    n = num
    numDigits = 0
    sum = 0
     
    # Find number of digits in num
    while n > 0:
        n //= 10
        numDigits += 1
     
    n = num
     
    # Calculate sum of digits raised to the power of numDigits
    while n > 0:
        digit = n % 10
        sum += math.pow(digit, numDigits)
        n //= 10
     
    # Check if num is Armstrong number or not
    if sum == num:
        return True
    return False

# Example 1
num1 = 1634
if isArmstrong(num1):
    print(num1, "is an Armstrong number.")
else:
    print(num1, "is not an Armstrong number.")
 
# Example 2
num2 = 120
if isArmstrong(num2):
    print(num2, "is an Armstrong number.")
else:
    print(num2, "is not an Armstrong number.")

1634 is an Armstrong number.
120 is not an Armstrong number.


Python program to check if a number is an Armstrong number in return statement

In [46]:
def is_armstrong_number(number):
    return sum(int(digit)**len(str(number)) for digit in str(number)) == number
 
 
# Example usage:
num = 153
if is_armstrong_number(num):
    print(f"{num} is an Armstrong number")
else:
    print(f"{num} is not an Armstrong number")

TypeError: 'float' object is not callable

## Question 7. Python Program to Find Area of a Circle

In [47]:
# Python program to find Area of a circle
 
def findArea(r):
    PI = 3.142
    return PI * (r*r);
 
# Driver method
print("Area is %.6f" % findArea(5));

Area is 78.550000


Python Program to Find Area of a Circle With Math library

In [48]:
# Python program to find Area of a circle
# using inbuild library
 
import math
def area(r):
  area = math.pi* pow(r,2)
  return print('Area of circle is:' ,area)
area(4)

Area of circle is: 50.26548245743669


## Qustion 8. Python Program to Print all Prime numbers in an Interval

Definition: A prime number is a natural number greater than 1 that has no positive divisors other than 1 and itself. The first few prime numbers are {2, 3, 5, 7, 11, ….}.

In [49]:
# Loop - https://www.geeksforgeeks.org/loops-in-python/

In [50]:
def prime(x, y):
    prime_list = []
    for i in range(x, y):
        if i == 0 or i == 1:
            continue
        else:
            for j in range(2, int(i/2)+1):
                if i % j == 0:
                    break
            else:
                prime_list.append(i)
    return prime_list
 
# Driver program
starting_range = 2
ending_range = 7
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are: ", lst)

The prime numbers in this range are:  [2, 3, 5]


Optimized way to find a Prime Number

The idea to solve this problem is to iterate the value from start to end using a for loop and for every number if it is divisible by any number except 1 and itself (prime number definition) then the flag variable will become 1. And in the next block if the flag value is zero then the only element will be appended to the list.

In [52]:
#For Loop - https://www.geeksforgeeks.org/python-for-loops/

In [53]:
# append() and extend() - https://www.geeksforgeeks.org/append-extend-python/

In [54]:
# Python program to print all
# prime number in an interval
 
def prime(starting_range, ending_range):
  lst=[]
  flag=0                   #Declaring flag variable
  for i in range(starting_range, ending_range):#elements range between starting and ending range
    for j in range(2,i): 
      if(i%j==0):     #checking if number is divisible or not
        flag=1        #if number is divisible, then flag variable will become 1
        break
      else:
        flag=0     
    if(flag==0):    #if flag variable is 0, then element will append in list 
      lst.append(i)
  return lst
 
# Driver program
starting_range = 2
ending_range = 7
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are: ", lst)

The prime numbers in this range are:  [2, 3, 5]


Sieve of Eratosthenes Method:

The sieve of Eratosthenes is one of the most efficient ways to find all primes smaller than N when N is smaller than 10 million or so.

Steps and implementation:

1. Create a boolean array prime[srt to n] and initialize all its entries as true.
2. Mark prime[0] and prime[1] as false because they are not prime numbers.
3. Starting from p = srt, for each prime number p less than or equal to the square root of n, mark all multiples of p greater than or equal to p*p as composite by setting prime[i] to false.
4. Finally, print all prime numbers between srt and n.

In [55]:
# Python program to find the prime numbers
# between a given interval using Sieve of Eratosthenes
import math
 
def SieveOfEratosthenes(srt, n):
    # Create a boolean array "prime[srt to n]" and
    # initialize all entries it as true. A value in
    # prime[i] will finally be false if i is Not a prime,
    # else true.
    prime = [True for i in range(n + 2 - srt)]
    prime[0] = False
    prime[1] = False
 
    for p in range(srt, int(math.sqrt(n))+1):
        # If prime[p] is not changed, then it is a prime
        if prime[p] == True:
            # Update all multiples of p greater than or
            # equal to the square of it numbers which are
            # multiple of p and are less than p^2 are
            # already been marked.
            for i in range(p*p, n+1, p):
                prime[i] = False
 
    # Print all prime numbers
    for p in range(srt, n+1):
        if prime[p]:
            print(p, end=" ")
 
# Driver Code
if __name__ == "__main__":
    srt = 1
    end = 10
    SieveOfEratosthenes(srt, end)

2 3 5 7 

Optimized Sieve of Eratosthenes

In [56]:
# Bitwise Sieve method - https://www.geeksforgeeks.org/bitwise-sieve/

One optimization of the above method is, we have skipped all even numbers altogether.  We reduce the size of the prime array to half. We also reduce all iterations to half.

In [57]:
# Python program to Print all Prime numbers in an Interval
# using simple optimized Sieve of Eratosthenes
# to reduce size of prime array to half and
# reducing iterations.
def bitwiseSieve(srt, n):
 
    # prime[i] is going to store
    # true if if i*2 + 1 is composite.
    prime = [0]*int(n / 2);
 
    # 2 is the only even prime so
    # we can ignore that. Loop
    # starts from 3.
    if (srt%2 == 0):
        srt += 1
    if(srt <= 2):
        srt = 3
    i = srt
    while(i * i < n):
        # If i is prime, mark all its
        # multiples as composite
        if (prime[int(i / 2)] == 0):
            j = i * i;
            while(j < n):
                prime[int(j / 2)] = 1;
                j += i * 2;
        i += 2;
 
    # writing 2 separately
    print(2,end=" ");
 
    # Printing other primes
    i = 3;
    while(i < n):
        if (prime[int(i / 2)] == 0):
            print(i,end=" ");
        i += 2;
 
 
# Driver code
if __name__=='__main__':
    srt = 1
    end = 10
    bitwiseSieve(srt, end)
 

2 3 5 7 

## Question 9. Python Program to Check Prime Number

The idea to solve this problem is to iterate through all the numbers starting from 2 to (N/2) using a for loop and for every number check if it divides N. If we find any number that divides, we return false. If we did not find any number between 2 and N/2 which divides N then it means that N is prime and we will return True.

In [58]:
num = 11
# If given number is greater than 1
if num > 1:
    # Iterate from 2 to n / 2
    for i in range(2, int(num/2)+1):
        # If num is divisible by any number between
        # 2 and n / 2, it is not prime
        if (num % i) == 0:
            print(num, "is not a prime number")
            break
    else:
        print(num, "is a prime number")
else:
    print(num, "is not a prime number")

11 is a prime number


Find Prime Numbers with a flag variable

Instead of checking till n, we can check till √n because a larger factor of n must be a multiple of a smaller factor that has been already checked. Now let’s see the code for the first optimization method ( i.e. checking till √n )

Check Prime Numbers Using Recursion

In [59]:
from math import sqrt
 
def Prime(number,itr):  #prime function to check given number prime or not
  if itr == 1:   #base condition
    return True
  if number % itr == 0:  #if given number divided by itr or not
    return False
  if Prime(number,itr-1) == False:   #Recursive function Call
    return False
     
  return True
  
num = 13
 
itr = int(sqrt(num)+1)
 
print(Prime(num,itr))

True


Check the Prime Trial Division Method

In [60]:
def is_prime_trial_division(n):
    # Check if the number is less than
    # or equal to 1, return False if it is
    if n <= 1:
        return False
    # Loop through all numbers from 2 to
    # the square root of n (rounded down to the nearest integer)
    for i in range(2, int(n**0.5)+1):
        # If n is divisible by any of these numbers, return False
        if n % i == 0:
            return False
    # If n is not divisible by any of these numbers, return True
    return True
 
 
# Test the function with n = 11
print(is_prime_trial_division(11))

True


Python Program to Check Prime Number Using a while loop to check divisibility

Initialize a variable i to 2, While i squared is less than or equal to n, check if n is divisible by i. If n is divisible by i, return False. Otherwise, increment i by 1. If the loop finishes without finding a divisor, return True.

In [61]:
import math
 
def is_prime(n):
    if n < 2:
        return False
    i = 2
    while i*i <= n:
        if n % i == 0:
            return False
        i += 1
    return True
 
print(is_prime(11))  # True
print(is_prime(1))   # False

True
False


Python Program to Check Prime Number Using Math module

The code implements a basic approach to check if a number is prime or not, by traversing all the numbers from 2 to sqrt(n)+1 and checking if n is divisible by any of those numbers. 

In [62]:
import math
 
def is_prime(n):
    if n <= 1:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True
 
n = 11
print(is_prime(n))

True


Python Program to Check Prime Number Using sympy.isprime() method

In the sympy module, we can test whether a given number n is prime or not using sympy.isprime() function. For n < 264 the answer is definitive; larger n values have a small probability of actually being pseudoprimes.

In [63]:
# Python program to check prime number
# using sympy.isprime() method
 
# importing sympy module
from sympy import *
 
# calling isprime function on different numbers
geek1 = isprime(30)
geek2 = isprime(13)
geek3 = isprime(2)
 
print(geek1) # check for 30 is prime or not
print(geek2) # check for 13 is prime or not
print(geek3) # check for 2 is prime or not

False
True
True


## Question 10. Python Program for n-th Fibonacci number

Python Program for n-th Fibonacci number Using recursion
Here we will use recursion function. The code defines a function Fibonacci(n) that calculates the nth Fibonacci number recursively. It checks for invalid input and returns the Fibonacci number based on the base cases (0 and 1) or by recursively calling itself with reduced values of n. The driver program prints the 10th Fibonacci number.

In [64]:
# Function for nth Fibonacci number
 
def Fibonacci(n):
    if n<= 0:
        print("Incorrect input")
    # First Fibonacci number is 0
    elif n == 1:
        return 0
    # Second Fibonacci number is 1
    elif n == 2:
        return 1
    else:
        return Fibonacci(n-1)+Fibonacci(n-2)
 
# Driver Program
 
print(Fibonacci(10))

34


Python Program for n-th Fibonacci number Using Dynamic Programming 

The code defines a function fibonacci(n) that calculates the nth Fibonacci number using dynamic programming. It initializes a list FibArray with the first two Fibonacci numbers (0 and 1). The function checks if the Fibonacci number for n is already present in FibArray and returns it. Otherwise, it calculates the Fibonacci number recursively, stores it in FibArray for future use, and returns the calculated value. The driver program prints the 9th Fibonacci number using this approach.

In [65]:
# Function for nth fibonacci number - Dynamic Programming
# Taking 1st two fibonacci numbers as 0 and 1
 
FibArray = [0, 1]
 
def fibonacci(n):
    if n<0:
        print("Incorrect input")
    elif n<= len(FibArray):
        return FibArray[n-1]
    else:
        temp_fib = fibonacci(n-1)+fibonacci(n-2)
        FibArray.append(temp_fib)
        return temp_fib
 
# Driver Program
 
print(fibonacci(9))

21


n-th Fibonacci number Using Dynamic Programming with Space Optimization

In [66]:
# Function for nth fibonacci number - Space Optimisation
# Taking 1st two fibonacci numbers as 0 and 1
 
def fibonacci(n):
    a = 0
    b = 1
    if n < 0:
        print("Incorrect input")
    elif n == 0:
        return a
    elif n == 1:
        return b
    else:
        for i in range(2, n):
            c = a + b
            a = b
            b = c
        return b
 
# Driver Program
 
print(fibonacci(9))

21


Python Program for n-th Fibonacci number Using Using Arrays 

The code defines a function fibonacci(n) that calculates the nth Fibonacci number by creating an array data containing the Fibonacci sequence up to the nth number. It initializes data with the first two Fibonacci numbers (0 and 1) and then iteratively calculates the subsequent numbers. The function returns the nth number from data. The driver program prints the 9th Fibonacci number using this approach.

In [67]:
# creating an array in the function to find the
#nth number in fibonacci series. [0, 1, 1, ...]
def fibonacci(n):
    if n <= 0:
        return "Incorrect Output"
    data = [0, 1]
    if n > 2:
        for i in range(2, n):
            data.append(data[i-1] + data[i-2])
    return data[n-1]
 
# Driver Program
print(fibonacci(9))

21


In [68]:
# To find the n-th Fibonacci Number using formula
from math import sqrt
# import square-root method from math library
def nthFib(n):
    res = (((1+sqrt(5))**n)-((1-sqrt(5)))**n)/(2**n*sqrt(5))
    # compute the n-th fibonacci number
    print(int(res),'is',str(n)+'th fibonacci number')
    # format and print the number
     
# driver code
nthFib(12)

144 is 12th fibonacci number


Python Program for n-th Fibonacci number Using power of the matrix {{1, 1}, {1, 0}}

This is another O(n) that relies on the fact that if we n times multiply the matrix M = {{1,1},{1,0}} to itself (in other words calculate power(M, n)), then we get the (n+1)th Fibonacci number as the element at row and column (0, 0) in the resultant matrix.
The matrix representation gives the following closed expression for the Fibonacci numbers: 

In [69]:
# Helper function that multiplies
# 2 matrices F and M of size 2*2,
# and puts the multiplication
# result back to F[][]
 
# Helper function that calculates
# F[][] raise to the power n and
# puts the result in F[][]
# Note that this function is
# designed only for fib() and
# won't work as general
# power function
def fib(n):
    F = [[1, 1],
         [1, 0]]
    if (n == 0):
        return 0
    power(F, n - 1)
    return F[0][0]
 
 
def multiply(F, M):
  x = (F[0][0] * M[0][0] + F[0][1] * M[1][0])
  y = (F[0][0] * M[0][1] + F[0][1] * M[1][1])
  z = (F[1][0] * M[0][0] + F[1][1] * M[1][0])
  w = (F[1][0] * M[0][1] + F[1][1] * M[1][1])
  F[0][0] = x
  F[0][1] = y
  F[1][0] = z
  F[1][1] = w
 
 
def power(F, n):
    M = [[1, 1], [1, 0]]
    # n - 1 times multiply the
    # matrix to {{1,0},{0,1}}
    for i in range(2, n + 1):
        multiply(F, M)
 
 
# Driver Code
if __name__ == "__main__":
    n = 9
    print(fib(n))
 

34


In [70]:
# Nth Fibonacci Number using
# optimized power of the matrix method
 
# function that returns nth
# Fibonacci number
 
 
def fib(n):
    F = [[1, 1],
         [1, 0]]
    if (n == 0):
        return 0
    power(F, n - 1)
 
    return F[0][0]
 
 
def multiply(F, M):
    x = (F[0][0] * M[0][0] +
         F[0][1] * M[1][0])
    y = (F[0][0] * M[0][1] +
         F[0][1] * M[1][1])
    z = (F[1][0] * M[0][0] +
         F[1][1] * M[1][0])
    w = (F[1][0] * M[0][1] +
         F[1][1] * M[1][1])
 
    F[0][0] = x
    F[0][1] = y
    F[1][0] = z
    F[1][1] = w
 
# Optimized version of
# power() in method 6
 
 
def power(F, n):
    if(n == 0 or n == 1):
        return
    M = [[1, 1],
         [1, 0]]
 
    power(F, n // 2)
    multiply(F, F)
 
    if (n % 2 != 0):
        multiply(F, M)
 
 
# Driver Code
if __name__ == "__main__":
    n = 5
    print(fib(n))

5


## Question 11. Python Program for How to check if a given number is Fibonacci number?

Given a number \’n\’, how to check if n is a Fibonacci number. First few Fibonacci numbers are 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, .. 

In [71]:
# python program to check if x is a perfect square
import math
 
# A utility function that returns true if x is perfect square
 
 
def isPerfectSquare(x):
    s = int(math.sqrt(x))
    return s*s == x
 
# Returns true if n is a Fibonacci Number, else false
 
 
def isFibonacci(n):
 
    # n is Fibonacci if one of 5*n*n + 4 or 5*n*n - 4 or both
    # is a perfect square
    return isPerfectSquare(5*n*n + 4) or isPerfectSquare(5*n*n - 4)
 
 
# A utility function to test above functions
for i in range(1, 11):
    if (isFibonacci(i) == True):
        print(i, "is a Fibonacci Number")
    else:
        print(i, "is a not Fibonacci Number ")

1 is a Fibonacci Number
2 is a Fibonacci Number
3 is a Fibonacci Number
4 is a not Fibonacci Number 
5 is a Fibonacci Number
6 is a not Fibonacci Number 
7 is a not Fibonacci Number 
8 is a Fibonacci Number
9 is a not Fibonacci Number 
10 is a not Fibonacci Number 


## Queston 12. Python Program for nth multiple of a number in Fibonacci Series

Given two integers n and k. Find position the nth multiple of K in the Fibonacci series. 

In [72]:
#Fibonacci series - https://www.geeksforgeeks.org/program-for-nth-fibonacci-number/

In [73]:
# Python Program to find position of n\'th multiple
# of a number k in Fibonacci Series
 
def findPosition(k, n):
    f1 = 0
    f2 = 1
    i =2;
    while i!=0:
        f3 = f1 + f2;
        f1 = f2;
        f2 = f3;
 
        if f2%k == 0:
            return n*i
 
        i+=1
         
    return
 
 
# Multiple no.
n = 5;
# Number of whose multiple we are finding
k = 4;
 
print("Position of n\'th multiple of k in"
                "Fibonacci Series is", findPosition(k,n));

Position of n'th multiple of k inFibonacci Series is 30


Method: Iterative approach using a list to store Fibonacci series

- First, we generate the Fibonacci series till the nth term.
- Then, we check if the current term is a multiple of k or not.
- If it is a multiple of k, we increment our counter variable.
- Once the counter becomes equal to n, we return the position of the current term in the Fibonacci series.

Steps

- Define a function nth_fib_multiple that takes two integers n and k as input.
- Initialize a list called fibonacci with the first two terms of the Fibonacci series.
- Initialize a variable called counter with value 0.
- Iterate from the third term of the Fibonacci series till the end.
- Check if the current term is a multiple of k or not.
- If it is a multiple of k, increment the counter.
- Check if the counter is equal to n or not.
- If it is equal to n, return the position of the current term in the Fibonacci series.
-If the counter is not equal to n, append the current term to the fibonacci list and continue the loop.

In [74]:
def nth_fib_multiple(n, k):
    fibonacci = [0, 1]
    counter = 0
    for i in range(2, 10000):
        current = fibonacci[i-1] + fibonacci[i-2]
        if current % k == 0:
            counter += 1
            if counter == n:
                return i
        fibonacci.append(current)
         
# Example usage
print(nth_fib_multiple(3, 2)) # Output: 9

9


Approach: Recursive Fibonacci series with memoization.

- Define a recursive function fib(n) that returns the nth number in the Fibonacci series.
- Implement memoization to store the values of previously calculated Fibonacci numbers.
- Define a function fib_multiple(k, n) that finds the nth multiple of k in the Fibonacci series.
- Iterate through the Fibonacci series until the nth multiple of k is found.
- Return the position of the nth multiple of k in the Fibonacci series.

In [75]:
# Recursive function to find the nth number in the Fibonacci series
def fib(n, memo):
    if n == 1 or n == 2:
        return 1
    if memo[n] != 0:
        return memo[n]
    memo[n] = fib(n-1, memo) + fib(n-2, memo)
    return memo[n]
 
# Function to find the nth multiple of k in the Fibonacci series
def fib_multiple(k, n):
    memo = [0] * 100
    count = 0
    for i in range(1, 100):
        if fib(i, memo) % k == 0:
            count += 1
            if count == n:
                return i
    return -1
 
# Test the function with sample inputs
k = 2
n = 3
result = fib_multiple(k, n)
print(f"The {n}th multiple of {k} in the Fibonacci series is at position {result}")

The 3th multiple of 2 in the Fibonacci series is at position 9


## Question 13. Program to print ASCII Value of a character